### PR TITLE
KAFKA-3412: multiple asynchronous commits causes send failures

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by hachikuji** [comment]
Ping @guozhangwang @ewencp for review.

This patch changes the asynchronous commit behavior to try and force the send during the commitAsync() call itself. Prior to this patch, we simply queued the commit to be sent in poll(), which causes send failures because NetworkClient doesn't allow more than one request to be sent on each invocation of poll() and ConsumerNetworkClient simply fails the rest. Based on my testing, this makes the problem go away in practice and also has the nice side effect of allowing heartbeats to be sent in commitAsync().

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java [hunk 1]
```diff
@@ -345,6 +345,9 @@ public void onFailure(RuntimeException e) {
                 cb.onComplete(offsets, e);
             }
         });
+
+        // ensure commit has a chance to be transmitted (without blocking on its completion)
+        client.poll(0);
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: Do we expect things like `WakeupException` to be acceptable here? Should we use `client.quickPoll()` instead?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java [hunk 2]
```diff
@@ -693,7 +693,6 @@ public boolean matches(ClientRequest request) {
 
         AtomicBoolean success = new AtomicBoolean(false);
         coordinator.commitOffsetsAsync(Collections.singletonMap(tp, new OffsetAndMetadata(100L)), callback(success));
-        consumerClient.poll(0);
```
Comment (ewencp) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:696]: There are 3 more similar `consumerClient.poll(0)` calls. Any reason not to remove those as well?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java (hunk 1 above)
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java]: I debated on that. We allow WakeupException to be thrown from commitSync(), but maybe it's a little surprising to see it in commitAsync(). The other aspect of quickPoll() is that it doesn't execute scheduled tasks, so no heartbeats would be sent. But maybe we could expose task execution in an argument to quickPoll().

**Comment by ijuma** [comment]
Nice that the fix is so small. Would it be possible to write a test that would fail before and pass now?

**Comment by hachikuji** [comment]
@ijuma Might be possible in an integration test. I'll have a look.

**Comment by ijuma** [comment]
Thanks!

**Comment by 13h3r** [comment]
Is it any semantic changes here? Will commit be performed without explicit call of `poll`?

**Comment by ewencp** [comment]
@hachikuji Do you actually need an integration test? Doesn't the exception get triggered from `ConsumerNetworkClient` (before you hit any mocks in `ConsumerCoordinatorTest`) and then get propagated to the callback for `commitOffsetsAsync`? So shouldn't this be testable just by doing exactly what the bug report says in a test in `ConsumerCoordinatorTest`?

**Comment by hachikuji** [comment]
@ewencp It's a bit awkward to test this through `MockClient`, which is what all the coordinator tests are using. I'm not sure you can do it in a way that doesn't amount to asserting that poll() is called in commitAsync(). It's really the integration between `ConsumerNetworkClient` and `ConsumerCoordinator` that needs testing. I guess your concern is creeping build time (more than 11 minutes now)?

@13h3r This changes behavior in two ways: (1) the commit will be sent when commitAsync() is called, and (2) commitAsync() can trigger heartbeats. My guess is that (1) is probably how users think this API works anyway. I really can't see a case where a user would depend on commitAsync() _not_ sending the commit. For users who are doing the standard poll loop, the difference won't be noticeable. For (2), this won't affect externally observed behavior at all, except that users won't fall out of the group as quickly ;-).

**Comment by ewencp** [comment]
@hachikuji I am worried about creeping build time, but also just hoping that it was testable via unit tests because they are much easier to debug/fix if anything changes. If an integration test is the way to go given current test setup and the code, go for it. I only mentioned the unit test approach because that test seems to use real versions of `ConsumerNetworkClient` and `ConsumerCoordinator` and I thought the `MockClient` could return a successful indicator for the first commit request and then the `ConsumerNetworkClient` would take care of failing any subsequent requests and invoking the callback. But I didn't look that closely yet to understand any implementation challenges there, I trust you if you say it doesn't work :)

**Comment by hachikuji** [comment]
@ewencp You might be right. I had forgotten that we use `ConsumerNetworkClient` directly. Let me see what that looks like.

**Comment by hachikuji** [comment]
@ewencp I ended up doing the integration test. We'd need to change the send behavior of `MockClient` to actually test the case that this patch is trying to fix.

**Comment by ewencp** [comment]
LGTM

**Pull request closed without merging**